driver/src/orderbook/retrying_retrieval.rs

```diff
@@ -0,0 +1,222 @@
+//! This module implements a retrying orderbook reader that accounts that allows
+//! the orderbook retrieval to optimistically start before the batch is
+//! finalized and will retry if state invalidating events are emitted.
+
+#![allow(dead_code)]
+
+use crate::contracts::stablex_contract::batch_exchange;
+use crate::models::{AccountState, Order};
+use crate::orderbook::paginated_auction_data_reader::PaginatedAuctionDataReader;
+use anyhow::{anyhow, Result};
+use ethcontract::web3::types::BlockId;
+use ethcontract::{Address, BlockNumber, Event, EventData};
+use futures::channel::{mpsc, oneshot};
+use futures::future::{BoxFuture, FutureExt};
+use futures::stream::{BoxStream, StreamExt};
+use futures::{select, try_join};
+use std::convert::TryInto;
+
+/// A trait for querying contract data with futures and streams.
+#[cfg_attr(test, mockall::automock)]
+#[allow(clippy::needless_lifetimes)]
+pub trait StableXContractAsync: Sync {
+    /// Retrieves the batch for a specific block number.
+    fn batch_id_at_block<'a>(&'a self, block: BlockId) -> BoxFuture<'a, Result<u32>>;
+    /// Searches for the block number of the last block of the given batch. If
+    /// the batch has not yet been finalized, then the current block number is
+    /// returned.
+    fn last_block_for_batch<'a>(&'a self, batch_id: u32) -> BoxFuture<'a, Result<u64>>;
+    /// Retrieves the latest solution for the given batch index.
+    fn encoded_orders_paginated<'a>(
+        &'a self,
+        previous_page_user: Address,
+        previous_page_user_offset: u16,
+        page_size: u16,
+        block_number: BlockNumber,
+    ) -> BoxFuture<'a, Result<Vec<u8>>>;
+    /// Streams incomming events from the contract that may signal that the
+    /// orderbook retrieval must be retried as the data has become stale.
+    fn all_events<'a>(&'a self) -> BoxStream<'a, Result<Event<batch_exchange::Event>>>;
+}
+
+fn read_orderbook(
+    contract: &dyn StableXContractAsync,
+    batch_id: u32,
+    page_size: u16,
+) -> Result<(AccountState, Vec<Order>)> {
+    futures::executor::block_on(async {
+        // NOTE: Use a bounded channel with a 0 sized buffer so that the sender
+        //   (i.e. the future that is polling the incomming events) can only
+        //   send 1 retry at a time.
+        let (retry_tx, retry_rx) = mpsc::channel::<u64>(0);
```

Comment:
What are `tx` and `rx`?

Reply:
Transmit and receive. They are commonly used abbreviations.